KNOX-1462 - Migrate from Log4j 1.x to 2.x

zeroflag opened the pull request:
## What changes were proposed in this pull request?

Due to security vulnerabilities and lack of maintenance log4j1 is replaced to log4j2. The two versions are incompatible with each other internally and on API level as well. 

The following changes were added:

* Package, class and method names are replaced to reflect then API changes 
* There is no MDC in log4j2. It is called ThreadContext and it only allows storing key value pair (String,String). Hence the additional `attachContext()` calls.
* Config file syntax is different and there is no `PropertyConfigurator` in log4j2. Config files are replaced to XML files.
* Custom `Appenders` are replaced to plugins.

## How was this patch tested?

Checked maven dependencies:

```
$ mvn dependency:tree | grep log4j:log4j:jar
$ 
```

Checked log rotation:

```
$ ls -al logs/gateway-audit.log*
-rw-r--r--  1 amagyar  staff   7065 Sep 11  2021 logs/gateway-audit.log
-rw-r--r--  1 amagyar  staff  10643 Sep  9 15:24 logs/gateway-audit.log.2021-09-09
-rw-r--r--  1 amagyar  staff   9222 Sep 10  2021 logs/gateway-audit.log.2021-09-10

$ ls -al logs/gateway.log*
-rw-r--r--  1 amagyar  staff   2741 Sep 11  2021 logs/gateway.log
-rw-r--r--  1 amagyar  staff  18005 Sep  9 15:24 logs/gateway.log.2021-09-09
-rw-r--r--  1 amagyar  staff   1556 Sep 10  2021 logs/gateway.log.2021-09-10
```

Checked the gateway.log + gateway-audit.log formats as well as the logs of the ldap server and knoxcli.

pzampino commented on gateway-i18n-logging-log4j/src/main/java/org/apache/knox/gateway/i18n/messages/loggers/log4j/Log4jMessageLogger.java:
```diff
@@ -37,42 +31,15 @@
-        info = new LocationInfo( caller.getFileName(), caller.getClassName(), caller.getMethodName(), Integer.toString(caller.getLineNumber()) );
-    }
-    return info;
+    logger.log(toLevel(messageLevel), messageText, thrown);
```
Is all the other info automatically populated by the logger now?

zeroflag replied:
No, this doesn't fully work yet. There is no callAppenders in log4j2, I need to find an equivalent method.

zeroflag replied:
I added the location info and now it seems to be working correctly.

zeroflag commented:
cc: @smolnar82 @lmccay @moresandeep

risdenk reviewed (changes requested):
Overall changes look reasonable. I left some comments about minor things that I think need cleaning up.

A few meta suggestions/questions:
* The comment in the top of this PR should be in the commit message
* This looks like it completely removes slf4j - is that correct? would it make sense to have the slf4j->log4j2 bridge added for any dependency that might use slf4j for logging?
* It would be good to include a comparison of the different logs that come out before and after this change. Specifically:
** gateway.log
** gateway-audit.log - this one is crucial since it is where most of the drastic changes happened in the logging framework

pzampino reviewed (commented):
I'm still reviewing, but asking the questions I've had thus far...

pzampino commented on gateway-test/src/test/java/org/apache/knox/gateway/KnoxCliLdapFuncTestNegative.java:
```diff
@@ -252,12 +249,7 @@ public void testBadTopology() throws Exception {
-    } finally {
-      NoOpAppender.resetOriginalAppenders( before );
-    }
+    cli.run( args2 );
```
What happened to the appender handling? Is the behavior of the test the same?

pzampino commented on gateway-test/src/test/java/org/apache/knox/gateway/KnoxCliLdapFuncTestPositive.java:
```diff
@@ -237,12 +234,7 @@ public void testLDAPAuth() throws Exception {
-    } finally {
-      NoOpAppender.resetOriginalAppenders( before );
-    }
+    cli.run(args2);
```
What happened to the appender handling? Is the behavior of the test the same?

pzampino commented on gateway-test/src/test/java/org/apache/knox/gateway/KnoxCliSysBindTest.java:
```diff
@@ -240,12 +237,7 @@ public void testLDAPAuth() throws Exception {
-    } finally {
-      NoOpAppender.resetOriginalAppenders( before );
-    }
+    cli.run( args3 );
```
What happened to the appender handling? Is the behavior of the test the same?

pzampino commented on gateway-test/src/test/java/org/apache/knox/gateway/deploy/DeploymentFactoryFuncTest.java:
```diff
@@ -180,7 +177,6 @@ public void testInvalidGenericProviderDeploymentContributor() {
     topology.addProvider( provider );
 
-    Enumeration<Appender> appenders = NoOpAppender.setUpAndReturnOriginalAppenders();
     try {
```
What happened to the appender handling? Is the behavior of the test the same?

pzampino commented on gateway-util-common/src/main/java/org/apache/knox/gateway/audit/log4j/audit/Log4jAuditor.java:
```diff
@@ -52,8 +51,8 @@
   public Log4jAuditor( String loggerName, String componentName, String serviceName ) {
-    logger = Logger.getLogger( loggerName );
-    logger.setAdditivity( false );
+    logger = (Logger) LogManager.getLogger( loggerName );
```
Is this cast necessary?

zeroflag replied:
There is no setAdditivity method in `org.apache.logging.log4j.Logger`, only in `org.apache.logging.log4j.core.Logger`.

This is how the the Log4j's compatibility API does this as well:

https://github.com/apache/logging-log4j2/blob/f630e16ecf196b5a7097a224a7997801b819ddd0/log4j-1.2-api/src/main/java/org/apache/log4j/legacy/core/CategoryUtil.java#L60

pzampino commented on gateway-util-common/src/main/java/org/apache/knox/gateway/audit/log4j/layout/AuditLayout.java:
```diff
@@ -20,60 +20,80 @@
+    appendParameter( eventContextData.getValue( AuditConstants.MDC_RESOURCE_NAME_KEY ) );
+    appendParameter( eventContextData.getValue( AuditConstants.MDC_OUTCOME_KEY ) );
+    String message = event.getMessage().getFormattedMessage();
+    sb.append( "null".equals(message) ? Strings.EMPTY : message ).append( System.lineSeparator() );
```
message is never null?

zeroflag replied:
I added an extra null check.

zeroflag commented on gateway-provider-rewrite/src/test/java/org/apache/knox/gateway/filter/rewrite/api/UrlRewriteRulesDescriptorFactoryTest.java:
```diff
@@ -98,10 +99,10 @@ public void testLoadMissingFile() throws IOException {
   @Test
   public void testLoadEmptyFile() {
-    Logger logger = org.apache.log4j.LogManager.getLogger( "org.apache.commons.digester3.Digester" );
+    Logger logger = (Logger)LogManager.getLogger( "org.apache.commons.digester3.Digester" );
```
There is no setLevel/setAdditivity method in `org.apache.logging.log4j.Logger`, only in `org.apache.logging.log4j.core.Logger`.

This is how the the Log4j's compatibility API does this as well:

https://github.com/apache/logging-log4j2/blob/f630e16ecf196b5a7097a224a7997801b819ddd0/log4j-1.2-api/src/main/java/org/apache/log4j/legacy/core/CategoryUtil.java#L60

zeroflag commented on gateway-server/src/test/java/org/apache/knox/gateway/descriptor/xml/XmlGatewayDescriptorImporterTest.java:
```diff
@@ -153,7 +154,7 @@ public void testXmlGatewayDescriptorLoadInvalid() throws IOException {
     Reader reader = new StringReader( xml );
     // Keep the tests quiet.  Ignore the stack trace that ends up being written to System.out.
-    Logger logger = Logger.getLogger( "org.apache.commons.digester3.Digester" );
+    Logger logger = (Logger)LogManager.getLogger( "org.apache.commons.digester3.Digester" );
```
Same as above.

zeroflag commented on gateway-server/src/test/resources/logging.properties:
```diff
@@ -1,87 +0,0 @@
-# limitations under the License.
-##########################################################################
-
-############################################################
-#  Default Logging Configuration File
-#
-# You can use a different file by specifying a filename
-# with the java.util.logging.config.file system property.  
-# For example java -Djava.util.logging.config.file=myfile
-############################################################
```
I couldn't find any use of this. This is under test resources and tests passed without this.

zeroflag commented on gateway-util-common/src/main/java/org/apache/knox/gateway/audit/api/CorrelationService.java:
```diff
@@ -23,19 +23,6 @@
  */
 public interface CorrelationService {
-
-  /**
-   * The recommended protocol header name used to transmit the correlation context over the network.
-   */
-  String PROTOCOL_HEADER = "X-Correlation-Context";
```
We don't use the constant or the string anywhere within the project.

zeroflag commented on gateway-provider-identity-assertion-common/src/main/java/org/apache/knox/gateway/identityasserter/common/filter/AbstractIdentityAssertionFilter.java:
```diff
@@ -110,7 +111,9 @@ protected void continueChainAsPrincipal(HttpServletRequestWrapper request, Servl
-            auditService.getContext().setProxyUsername( mappedPrincipalName );
+            AuditContext context = auditService.getContext();
+            context.setProxyUsername( mappedPrincipalName );
+            auditService.attachContext(context);
```
In log4j1 the `MDC` held a reference to an object, so changing the `proxyUserName` on this object changed the `MDC` as well. But in log4j2 there is a `ThreadContext` instead of `MDC` which can hold only key value pairs. So changing the `proxyUserName` on the `AuditContext` won't change the the threadContext without reattaching (thus populating the key values) again.

pzampino replied:
Ack

zeroflag commented on pom.xml:
```diff
@@ -234,8 +234,7 @@
         <libpam4j.version>1.11</libpam4j.version>
-        <log4j.version>1.2.17</log4j.version>
-        <log4j2.version>2.14.0</log4j2.version>
+        <log4j.version>2.14.0</log4j.version>
```
Ok.

zeroflag commented on gateway-i18n-logging-log4j/src/main/java/org/apache/knox/gateway/i18n/messages/loggers/log4j/Log4jMessageLogger.java:
```diff
@@ -37,42 +33,15 @@
-        info = new LocationInfo( caller.getFileName(), caller.getClassName(), caller.getMethodName(), Integer.toString(caller.getLineNumber()) );
-    }
-    return info;
+    logger.logMessage(toLevel(messageLevel), null, CLASS_NAME, caller, new SimpleMessage(messageText), thrown);
```
@risdenk 

```
amagyar-MBPloaner:test amagyar$ tail ~/Downloads/1gateway.log 
2021-09-13 16:20:15,339 WARN  federation.jwt (AbstractJWTFilter.java:validateToken(339)) - Failed to verify the token signature of eyJraW...plfBXA (69d7e7ed...929fd3b25b82)
2021-09-13 16:20:19,495 INFO  knox.gateway (KnoxLdapRealm.java:getUserDn(688)) - Computed userDn: uid=admin,ou=people,dc=hadoop,dc=apache,dc=org using dnTemplate for principal: admin
2021-09-13 16:20:19,982 INFO  knox.gateway (CookieUtils.java:getCookiesForName(46)) - Unable to find cookie with name: original-url
2021-09-13 16:20:19,989 INFO  service.knoxsso (WebSSOResource.java:addJWTHadoopCookie(386)) - JWT cookie successfully added.
2021-09-13 16:20:19,990 INFO  service.knoxsso (WebSSOResource.java:getAuthenticationToken(278)) - About to redirect to original URL: https://localhost:8443/gateway/homepage/tokengen/index.html
2021-09-13 16:20:20,368 WARN  federation.jwt (SSOCookieFederationFilter.java:init(95)) - Configuration for authentication provider URL is missing - will derive default URL.
2021-09-13 16:20:20,467 INFO  service.knoxtoken (TokenResource.java:init(234)) - Server management of token state is enabled for the "homepage" topology.
2021-09-13 16:20:21,705 INFO  service.knoxtoken (TokenResource.java:getAuthenticationToken(673)) - Knox Token service (homepage) issued token eyJqa3...WK-log (d651a8ba...dee52d5d3433)
2021-09-13 16:20:30,610 INFO  token.state (AliasBasedTokenStateService.java:persistTokenState(243)) - Creating token state aliases
2021-09-13 16:20:30,915 INFO  token.state (AliasBasedTokenStateService.java:persistTokenState(252)) - Created token state aliases for d651a8ba...dee52d5d3433
amagyar-MBPloaner:test amagyar$ tail ~/Downloads/2gateway.log 
2021-09-13T16:10:27,612 INFO  knox.gateway (CookieUtils.java:getCookiesForName(46)) - Unable to find cookie with name: original-url
2021-09-13T16:10:27,619 INFO  service.knoxsso (WebSSOResource.java:addJWTHadoopCookie(386)) - JWT cookie successfully added.
2021-09-13T16:10:27,620 INFO  service.knoxsso (WebSSOResource.java:getAuthenticationToken(278)) - About to redirect to original URL: https://localhost:8443/gateway/homepage/tokengen/index.html
2021-09-13T16:10:27,825 WARN  federation.jwt (SSOCookieFederationFilter.java:init(95)) - Configuration for authentication provider URL is missing - will derive default URL.
2021-09-13T16:10:27,923 INFO  service.knoxtoken (TokenResource.java:init(234)) - Server management of token state is enabled for the "homepage" topology.
2021-09-13T16:10:32,237 INFO  service.knoxtoken (TokenResource.java:getAuthenticationToken(673)) - Knox Token service (homepage) issued token eyJqa3...XWUzNg (1ab3aafb...f2d5b0ba8466)
2021-09-13T16:10:32,634 INFO  service.knoxtoken (TokenResource.java:getAuthenticationToken(673)) - Knox Token service (homepage) issued token eyJqa3...voLZGw (dec74aac...96b76aa1b8b7)
2021-09-13T16:10:45,996 INFO  token.state (AliasBasedTokenStateService.java:persistTokenState(243)) - Creating token state aliases
2021-09-13T16:10:46,608 INFO  token.state (AliasBasedTokenStateService.java:persistTokenState(252)) - Created token state aliases for dec74aac...96b76aa1b8b7
2021-09-13T16:10:46,609 INFO  token.state (AliasBasedTokenStateService.java:persistTokenState(252)) - Created token state aliases for 1ab3aafb...f2d5b0ba8466
amagyar-MBPloaner:test amagyar$ 
```

I see one difference in the date time format there is a 'T' between the date and time in the log2j version. I'll check that why. Other than this it looks the same.

I see no difference between the gateway-audit logs:

```
amagyar-MBPloaner:test amagyar$ tail ~/Downloads/1gateway-audit.log
21/09/13 16:20:20 |||audit|[0:0:0:0:0:0:0:1]|tokengen|admin|||access|uri|/gateway/homepage/tokengen/index.html|success|Response status: 200
21/09/13 16:20:20 ||0362085a-ccdb-4baf-8ec7-35aa9258da94|audit|[0:0:0:0:0:0:0:1]|tokengen||||access|uri|/gateway/homepage/tokengen/index.html|unavailable|Request method: GET
21/09/13 16:20:20 ||0362085a-ccdb-4baf-8ec7-35aa9258da94|audit|[0:0:0:0:0:0:0:1]|tokengen|admin|||authentication|uri|/gateway/homepage/tokengen/index.html|success|
21/09/13 16:20:20 |||audit|[0:0:0:0:0:0:0:1]|tokengen|admin|||access|uri|/gateway/homepage/tokengen/index.html|success|Response status: 200
21/09/13 16:20:20 ||3fa7c4b8-4d1b-4d22-bb8f-ab4c5d84ed14|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN||||access|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|unavailable|Request method: GET
21/09/13 16:20:20 ||3fa7c4b8-4d1b-4d22-bb8f-ab4c5d84ed14|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||authentication|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|success|
21/09/13 16:20:20 |||audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||access|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|success|Response status: 200
21/09/13 16:20:21 ||099c85e8-1746-4017-804e-ce1c46914e69|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN||||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|unavailable|Request method: GET
21/09/13 16:20:21 ||099c85e8-1746-4017-804e-ce1c46914e69|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||authentication|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|
21/09/13 16:20:21 |||audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|Response status: 200
amagyar-MBPloaner:test amagyar$ tail ~/Downloads/2gateway-audit.log
21/09/13 14:10:27 |||audit|[0:0:0:0:0:0:0:1]|tokengen|admin|||access|uri|/gateway/homepage/tokengen/fonts/fontawesome/fontawesome-webfont.woff?v=3.2.1|success|Response status: 200
21/09/13 14:10:27 ||648f99cb-66ad-4206-84c6-2422684aec2b|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN||||access|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|unavailable|Request method: GET
21/09/13 14:10:27 ||648f99cb-66ad-4206-84c6-2422684aec2b|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||authentication|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|success|
21/09/13 14:10:27 |||audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||access|uri|/gateway/homepage/knoxtoken/api/v1/token/getTssStatus|success|Response status: 200
21/09/13 14:10:32 ||89e5d866-1423-47b6-9138-0e179356ec2a|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN||||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|unavailable|Request method: GET
21/09/13 14:10:32 ||89e5d866-1423-47b6-9138-0e179356ec2a|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||authentication|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|
21/09/13 14:10:32 |||audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|Response status: 200
21/09/13 14:10:32 ||fab4d745-8f70-4013-8394-a3da40796716|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN||||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|unavailable|Request method: GET
21/09/13 14:10:32 ||fab4d745-8f70-4013-8394-a3da40796716|audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||authentication|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|
21/09/13 14:10:32 |||audit|[0:0:0:0:0:0:0:1]|KNOXTOKEN|admin|||access|uri|/gateway/homepage/knoxtoken/api/v1/token?lifespan=P0DT1H0M|success|Response status: 200
amagyar-MBPloaner:test amagyar$ 
```

zeroflag replied:
In log4j1 ISO8601 datetime pattern doesn't contain the 'T' separator between the date and time but in log4j2 it is there. I replaced the pattern to `%d{yyyy-MM-dd' 'HH:mm:ss,SSS}` in the gateway's log4j config so that it looks the same with log4j2.

zeroflag commented on gateway-test/src/test/resources/log4j.properties:
```diff
@@ -1,44 +0,0 @@
-log4j.logger.audit = INFO, collectappender
-log4j.appender.collectappender = org.apache.knox.test.log.CollectAppender
-
-#log4j.logger.org.apache.knox.gateway=DEBUG
-#log4j.logger.org.apache.knox.test=DEBUG
-#log4j.logger.org.apache.knox.gateway.http=TRACE
-#log4j.logger.org.apache.knox.gateway.http.request.body=OFF
-#log4j.logger.org.apache.knox.gateway.http.response.body=OFF
-
-#log4j.logger.org.apache.directory=DEBUG
-#log4j.logger.org.eclipse.jetty=DEBUG
-#log4j.logger.org.apache.shiro=DEBUG
-#log4j.logger.org.apache.http=DEBUG
-#log4j.logger.org.apache.http.headers=DEBUG
-#log4j.logger.org.apache.http.wire=DEBUG
-#log4j.logger.org.apache.http.client=DEBUG
-
-# Hide verbose logging - see KNOX-1718 for more details
```
I put it back to the XML.

zeroflag commented on gateway-provider-rewrite/src/test/resources/log4j.properties:
```diff
@@ -1,37 +0,0 @@
-log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
-log4j.appender.stdout.layout.ConversionPattern=%5p [%c] %m%n
-
-#log4j.logger.org.apache.knox.gateway=DEBUG
-#log4j.logger.org.apache.knox.gateway.http=TRACE
-#log4j.logger.org.apache.knox.gateway.http.request.body=OFF
-#log4j.logger.org.apache.knox.gateway.http.response.body=OFF
-
-#log4j.logger.org.apache.directory=DEBUG
-#log4j.logger.org.eclipse.jetty=DEBUG
-#log4j.logger.org.apache.shiro=DEBUG
-#log4j.logger.org.apache.http=DEBUG
-#log4j.logger.org.apache.http.headers=DEBUG
-#log4j.logger.org.apache.http.wire=DEBUG
-#log4j.logger.org.apache.http.client=DEBUG
```
I put it back to the XML.

zeroflag commented on gateway-release/home/conf/gateway-log4j.properties:
```diff
@@ -1,85 +0,0 @@
-log4j.appender.auditfile.DatePattern = '.'yyyy-MM-dd
-log4j.appender.auditfile.layout = org.apache.knox.gateway.audit.log4j.layout.AuditLayout
-
-#log4j.logger.org.apache.knox.gateway.access=TRACE,httpaccess
-#log4j.additivity.org.apache.knox.gateway.access=false
-
-#log4j.logger.org.apache.knox.gateway.http=TRACE,httpserver
-#log4j.additivity.org.apache.knox.gateway.http=false
-##log4j.logger.org.apache.knox.gateway.http.request.headers=OFF
-##log4j.logger.org.apache.knox.gateway.http.response.headers=OFF
-##log4j.logger.org.apache.knox.gateway.http.request.body=OFF
-##log4j.logger.org.apache.knox.gateway.http.response.body=OFF
-
-#log4j.logger.org.apache.http.wire=DEBUG,httpclient
-#log4j.additivity.org.apache.http.wire=false
-
-#log4j.appender.httpaccess=org.apache.log4j.DailyRollingFileAppender
-#log4j.appender.httpaccess.File=${app.log.dir}/${launcher.name}-http-access.log
-#log4j.appender.httpaccess.DatePattern=.yyyy-MM-dd
-#log4j.appender.httpaccess.layout=org.apache.log4j.PatternLayout
-#log4j.appender.httpaccess.layout.ConversionPattern=%d{ISO8601}|%t|%m%n
-
-#log4j.appender.httpserver=org.apache.log4j.DailyRollingFileAppender
-#log4j.appender.httpserver.File=${app.log.dir}/${launcher.name}-http-server.log
-#log4j.appender.httpserver.DatePattern=.yyyy-MM-dd
-#log4j.appender.httpserver.layout=org.apache.log4j.PatternLayout
-#log4j.appender.httpserver.layout.ConversionPattern=%d{ISO8601}|%t|%m%n
-
-#log4j.appender.httpclient=org.apache.log4j.DailyRollingFileAppender
-#log4j.appender.httpclient.File=${app.log.dir}/${launcher.name}-http-client.log
-#log4j.appender.httpclient.DatePattern=.yyyy-MM-dd
-#log4j.appender.httpclient.layout=org.apache.log4j.PatternLayout
-#log4j.appender.httpclient.layout.ConversionPattern=%d{ISO8601}|%t|%m%n
-
-# Apache Shiro Related logging - KNOX-757
-#log4j.logger.org.springframework=DEBUG
-#log4j.logger.net.sf.ehcache=DEBUG
-#log4j.logger.org.apache.shiro.util.ThreadContext=DEBUG
```
I put it back to the XML.

zeroflag commented on gateway-server/src/test/resources/log4j.properties:
```diff
@@ -1,35 +0,0 @@
-log4j.logger.audit = INFO, collectappender
-log4j.appender.collectappender = org.apache.knox.test.log.CollectAppender
-
-#log4j.logger.org.apache.knox.gateway=DEBUG
-#log4j.logger.org.eclipse.jetty=DEBUG
-#log4j.logger.org.apache.shiro=DEBUG
-#log4j.logger.org.apache.http=DEBUG
-#log4j.logger.org.apache.http.wire=DEBUG
-#log4j.logger.org.apache.http.client=DEBUG
```
I put it back to the XML.

zeroflag commented on gateway-provider-rewrite/src/test/java/org/apache/knox/gateway/filter/rewrite/api/UrlRewriteServletFilterTest.java:
```diff
@@ -527,41 +523,35 @@ public void testRequestXmlBodyRewriteWithFilterInitParam() throws Exception {
-    } finally {
-      NoOpAppender.resetOriginalAppenders( realAppenders );
-    }
+    Map<String,String> initParams = new HashMap<>();
```
I'm not quite sure what was the purpose of the NoOpAppender. It just temporary disabled logging while the test was running. The test doesn't check any log related things. The class was added in 2013 but the commit message doesn't contain any Jira reference so I couldn't get any info on the intent. I'll try find out out why it was added but so far it looks like it doesn't influence the test run.

zeroflag commented on gateway-provider-ha/src/test/java/org/apache/knox/gateway/ha/provider/impl/AtlasZookeeperURLManagerTest.java:
```diff
@@ -47,6 +47,7 @@
 
     @Before
     public void setUp() throws Exception {
+        System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-provider-ha/src/test/java/org/apache/knox/gateway/ha/provider/impl/HBaseZookeeperURLManagerTest.java:
```diff
@@ -49,6 +49,7 @@
 
   @Before
   public void setUp() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-provider-ha/src/test/java/org/apache/knox/gateway/ha/provider/impl/HS2ZookeeperURLManagerTest.java:
```diff
@@ -43,6 +43,7 @@
 
   @Before
   public void setUp() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-test/src/test/java/org/apache/knox/gateway/topology/monitor/RemoteConfigurationMonitorTest.java:
```diff
@@ -155,6 +155,8 @@ private static File setupDigestSaslConfig(String username, String password) thro
      * Configure and start the ZooKeeper test cluster, and create the znodes monitored by the RemoteConfigurationMonitor.
      */
     private void configureAndStartZKCluster() throws Exception {
+        System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-provider-ha/src/test/java/org/apache/knox/gateway/ha/provider/impl/KafkaZookeeperURLManagerTest.java:
```diff
@@ -44,6 +44,7 @@
 
   @Before
   public void setUp() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-provider-ha/src/test/java/org/apache/knox/gateway/ha/provider/impl/SOLRZookeeperURLManagerTest.java:
```diff
@@ -48,6 +48,7 @@
 
   @Before
   public void setUp() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-server/src/test/java/org/apache/knox/gateway/services/security/impl/ZookeeperRemoteAliasMonitorTest.java:
```diff
@@ -69,6 +69,8 @@
 
   @BeforeClass
   public static void setupSuite() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-server/src/test/java/org/apache/knox/gateway/services/security/impl/ZookeeperRemoteAliasServiceTest.java:
```diff
@@ -96,6 +96,8 @@ public static void setupSuite() throws Exception {
   }
 
   private static void configureAndStartZKCluster() throws Exception {
+    System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-server/src/test/java/org/apache/knox/gateway/topology/monitor/ZooKeeperConfigurationMonitorTest.java:
```diff
@@ -85,6 +85,8 @@ public static void setupSuite() throws Exception {
     }
 
     private static void configureAndStartZKCluster() throws Exception {
+        System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented on gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryClientServiceTestBase.java:
```diff
@@ -54,6 +54,8 @@
     protected TestingCluster setupAndStartSecureTestZooKeeper(String principal, String digestPassword) throws Exception {
         final boolean applyAuthentication = (principal != null);
 
+        System.setProperty("zookeeper.jmx.log4j.disable", "true");
```
I removed these, they're not needed.

zeroflag commented:
> * This looks like it completely removes slf4j - is that correct? would it make sense to have the slf4j->log4j2 bridge added for any dependency that might use slf4j for logging?

@risdenk Thanks for going through this big patch. 

At a few places in the Knox code we used `org.slf4j.LoggerFactory` and `org.slf4j.Logger` directly. I replaced these with Log4j2 and removed the slf4j dependency to make sure we won't use these again. 

If there is a third party dependency that uses slf4j for logging, that still should work since we have `log4j-slf4j-impl-2.14.0.jar` on the class path.

```
$ ls -al  dep/ | grep log4j
-rw-r--r--    1 amagyar  staff   301418 Jan 22  2020 log4j-api-2.14.0.jar
-rw-r--r--    1 amagyar  staff  1762731 Jan 22  2020 log4j-core-2.14.0.jar
-rw-r--r--    1 amagyar  staff    23871 Jan 22  2020 log4j-slf4j-impl-2.14.0.jar
```

pzampino reviewed (approved):
LGTM

risdenk reviewed (approved):
All in all the code changes look good.

I didn't see any documentation updates - https://knox.apache.org/books/knox-1-5-0/user-guide.html - At least on this page there are a bunch of log4j references that are specific to log4j 1.x.

The other concern I have is upgrades. If this gets merged - and released in the next 1.x release of Apache Knox - what happens when a user upgrades Knox? All the logging needs to be reconfigured? Is there a migration path?
